drv/lpc55-swd/src/main.rs

```diff
@@ -1058,21 +1358,588 @@ impl ServerImpl {
+                ));
+                return Err(SpCtrlError::Fault);
+            }
+            if deadline <= sys_get_timer().now {
+                // If a human is holding down a physical reset button then
+                // SP_RESET may have never been released.
+                let gpio = Pins::from(self.gpio);
+                let sp_reset_state = gpio.read_val(ROT_TO_SP_RESET_L_IN);
+                ringbuf_entry!(Trace::DidNotHalt { sp_reset_state });
+                break Err(SpCtrlError::Timeout);
+            }
+            hl::sleep_for(1);
+        }
+    }
+
+    fn do_resume(&mut self) -> Result<(), SpCtrlError> {
+        if self.dp_write_bitflags::<Dhcsr>(Dhcsr::resume()).is_ok() {
+            ringbuf_entry!(Trace::Resumed);
+            Ok(())
+        } else {
+            ringbuf_entry!(Trace::ResumeFail);
+            Err(SpCtrlError::Fault)
+        }
+    }
+
+    fn invalidate_sp_measurement(&mut self) -> Result<(), AttestError> {
+        match self.attest.reset() {
+            Ok(()) => {
+                ringbuf_entry!(Trace::InvalidatedSpMeasurement);
+                Ok(())
+            }
+            Err(e) => {
+                ringbuf_entry!(Trace::AttestError(e));
+                Err(e)
+            }
+        }
+    }
+
+    // Return true if necessary work was done.
+    // Return false if any current SP measurement should be invalidated.
+    fn do_handle_sp_reset(&mut self) -> bool {
+        let start = sys_get_timer().now;
+        let gpio = Pins::from(self.gpio);
+        const SLOT: PintSlot = ROT_TO_SP_RESET_L_IN_PINT_SLOT;
+        let mut need_undo = Undo::from_bits_retain(0);
+
+        // Did SP_RESET transition to Zero?
+        if let Ok(Some(detected)) =
+            gpio.pint_op(SLOT, PintOp::Detected, PintCondition::Falling)
+        {
+            if !detected {
+                // Use of sys_irq_control_clear_pending(...) should avoid
+                // appearance of a "spurious" interrupt.
+                // Otherwise, cases where we assert SP_RESET then clean-up the PINT
+                // condition will have a pending notification.
+                ringbuf_entry!(Trace::SpResetNotAsserted);
+                return true; // no work required.
+            }
+        } else {
+            // The pint_op parameters are for a configured PINT slot for one of
+            // our configured GPIO pins. We're testing a status bit in a register.
+            unreachable!();
+        }
+
+        // Not ok yet: A reset happened. If we don't get a measurement then
+        // make sure that the old one is invalidated.
+        let mut error = false;
+
+        ringbuf_entry!(Trace::SpResetAsserted);
+
+        // This notification handler should be compatible with watchdog but
+        // will result in the invalidation of any dumps held in the SP if successful.
+
+        // TODO: confirm that bank-flipping SP update watchdog is working.
```

Comment:
Can we remove this TODO?

Reply:
I'll re-run my fault injection test to make sure all is well then remove the comment.